sympy/functions/combinatorial/factorials.py

```diff
@@ -740,8 +740,11 @@ def eval(cls, x, k):
                                             range(1, abs(int(k)) + 1), 1)
 
     def _eval_rewrite_as_gamma(self, x, k, **kwargs):
-        from sympy import gamma
-        return (-1)**k*gamma(k - x) / gamma(-x)
+        from sympy import gamma, Piecewise
+
+        return Piecewise(
```

Comment:
I second @smichr , returning a Piecewise isn't necessarily a good idea.
return a rewrite only when the assumptions are valid.
in this case,
- if both n and k are integers
  - if n is negative
    - (-1)**k*gamma(k - n) / gamma(-n)
  - if n is positive
    - gamma(n + 1)/gamma(-k + n + 1)
- elif any(i.integers is False for i in (n, k))
  - gamma(n + 1)/gamma(-k + n + 1)
or some version like this, just make sure the assumptions hold

Reply:
He suggested this itself here: https://github.com/sympy/sympy/pull/18696#discussion_r420329804

But it is causing some doctests to fail: https://github.com/sympy/sympy/pull/18696#discussion_r421946373

Reply:
Yeah fix those, give the correct assumptions.

Reply:
Do you mean we should edit the doctests?

Reply:
Check out 
https://github.com/sympy/sympy/pull/18960/files?short_path=6f21732#diff-6f217322f241daeecdd5a7605fd45b03

Reply:
Yeah, actually I was thinking similarly. 
Ping @jksuom @smichr. Should we edit the assumptions in the rst doctests?

Reply:
imo Not before you get the implementation correct tho.

Reply:
Yeah as I have already mentioned except the rst doctests everything else was working fine, so the implementation seems correct.